File: java-algorithm/src/main/java/com/leetcode/cn/Lcof.java
package com.leetcode.cn;

import java.util.ArrayList;
import java.util.List;

public class Lcof {

    public int translateNum(int num) {
        String str = String.valueOf(num);
        int length = str.length();
        List<Integer> store = new ArrayList<>(length + 1);
        for (int i = 0; i < length + 1; i++) {
            store.add(1);
        }

        for (int i = length - 2; i >= 0; i--) {
            // f(n) = f(n+1) + (if exists ? f(n+2) : 0)
            store.set(i, store.get(i + 1));
            int temp = Integer.parseInt(str.substring(i, i + 2));
            if (temp < 26 && temp > 9) {
                store.set(i, store.get(i + 1) + store.get(i + 2));
            }
        }

        return store.get(0);
    }

}

File: python-algorithm/other/leetcode_cn/lcof.py
"""面试题46. 把数字翻译成字符串
https://leetcode-cn.com/problems/ba-shu-zi-fan-yi-cheng-zi-fu-chuan-lcof/

给定一个数字，我们按照如下规则把它翻译为字符串：0 翻译成 “a” ，1 翻译成 “b”，……，11 翻译成 “l”，……，25 翻译成 “z”。
一个数字可能有多个翻译。请编程实现一个函数，用来计算一个数字有多少种不同的翻译方法。

示例 1:

  输入: 12258
  输出: 5
  解释: 12258有5种不同的翻译，分别是"bccfi", "bwfi", "bczi", "mcfi"和"mzi"

提示：
  0 <= num < 231

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/ba-shu-zi-fan-yi-cheng-zi-fu-chuan-lcof
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
"""


class Solution:
    def translate_num(self, num: int) -> int:
        num_str = str(num)
        size = len(num_str)
        store = [1] * (size + 1)
        for i in range(size - 2, -1, -1):
            store[i] = store[i + 1]
            temp = int(num_str[i: i + 2])
            if 26 > temp > 9:
                store[i] = store[i] + store[i + 2]
        return store[0]
